Get country code by location ID for get campaigns API

ianlin opened the pull request:
### Changes proposed in this Pull Request:

Part of #1248 and an enhancement for #1280. This PR integrates the work in #1282 to convert `geo_target_constant` in [LocationInfo](https://developers.google.com/google-ads/api/reference/rpc/v9/LocationInfo) (location ID) to country code in ISO 3166-1 alpha-2 format.

Also in this PR it refactors the logic of filtering out the empty geo target constants. Previously it queried lots of campaign criteria with empty geo target constants and filter it out in the application code. The new approach suggested in https://github.com/woocommerce/google-listings-and-ads/pull/1280#discussion_r820747160 is adding the logic as part of [Google Ads Query Language](https://developers.google.com/google-ads/api/docs/query/overview) by using `campaign_criterion.location.geo_target_constant IS NOT NULL`. This reduce the amount of data returned by Google Ads API and also enhance the code readability.

### Screenshots

#### GET ads/camapigns

<img width="1362" alt="Screenshot 2022-03-09 at 17 21 47" src="https://user-images.githubusercontent.com/914406/157411704-583d7531-2e91-45f5-8298-3004321b391b.png">

#### GET ads/camapigns/:campaign-id

<img width="1364" alt="Screenshot 2022-03-09 at 17 21 26" src="https://user-images.githubusercontent.com/914406/157411767-1d91afbf-bd15-4752-8dca-fcb01f90bf2a.png">

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Create some paid campaigns in GLA's dashboard: `/wp-admin/admin.php?page=wc-admin&path=/google/dashboard`.
2. Go to Google Ads Campaigns Dashboard and add targeted locations in some of the campaigns you created.
    - <img width="2555" alt="Screenshot 2022-03-03 at 13 09 34" src="https://user-images.githubusercontent.com/914406/156500476-8d9e00d5-ff0f-429b-895c-3a4ffedae06d.png">
3. Using Postman, request to `GET ads/campaigns` and `GET ads/campaigns/:campaign-id` like the screenshots in the `Screenshots` section. Remember to add your `wp-cookie` and `x-wp-nonce` in the request headers. 
    - To get the value of `wp-cookie` and `x-wp-nonce`, go to `Marketing -> Google Listings & Ads`, open Chrome dev tool's Network tab, clicks any requests to WP, e.g. `wc/gla/ads/campaigns`, you can get those values in the request headers.

### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Update - Get country code by location ID for get campaigns API

ianlin committed:
Get campaign criterion with non-null geo constants by Google's query

ianlin committed:
Respond country code instead of geo location ID for GET ads/campaign(s)

By using GoogleHelper's `find_country_code_by_id()` method.

mikkamp commented on src/API/Google/AdsCampaign.php:
```diff
@@ -8,6 +8,7 @@
 use Automattic\WooCommerce\GoogleListingsAndAds\API\MicroTrait;
 use Automattic\WooCommerce\GoogleListingsAndAds\Exception\ExceptionWithResponseData;
 use Automattic\WooCommerce\GoogleListingsAndAds\Exception\InvalidQuery;
```
From the previous PR, but InvalidQuery is no longer used.

ianlin replied:
Updated in 0f7e523, thanks!

mikkamp commented on src/API/Google/AdsCampaign.php:
```diff
@@ -415,26 +417,27 @@ protected function combine_campaigns_and_campaign_criterion_results( array $camp
+			$geo_target_constant = $location->getGeoTargetConstant();
+			$location_id         = $this->parse_geo_target_location_id( $geo_target_constant );
+
+			$google_helper = $this->container->get( GoogleHelper::class );
```
Have mixed feelings on whether this should be retrieved from the container or be added as a direct dependency. But for sure we don't need to fetch this class within the for loop. Also if it's fetched from the container then it's helpful to add it to the list at the top:

```
 * ContainerAware used for:
 * - AdsGroup
```

ianlin replied:
I've got the mutual feelings and I found there was a discussion here: `pb0Spc-1u4-p2`. From the conclusion there I think it's pretty clear that we should use direct dependency as much as possible, and in this case there were only two dependencies for the AdsCampaign class so it's still OK to add one more (google_helper) as its dependency. Thanks for raising this question! Updated in 0247998.

mikkamp commented on src/Internal/DependencyManagement/GoogleServiceProvider.php:
```diff
@@ -100,7 +101,7 @@ public function register() {
 
 		$this->share( Ads::class, GoogleAdsClient::class );
-		$this->share( AdsCampaign::class, GoogleAdsClient::class, AdsCampaignBudget::class );
+		$this->share( AdsCampaign::class, GoogleAdsClient::class, AdsCampaignBudget::class, GoogleHelper::class );
```
The AdsCampaign constructor only accepts the GoogleAdsClient and AdsCampaignBudget parameters. So adding GoogleHelper here does nothing (it's fetched from the container within the class).

ianlin replied:
Thanks for catching this! While I've updated to inject google_helper as AdsCampaign's dependency, this will remain unchanged.

mikkamp commented on src/API/Google/AdsCampaign.php:
```diff
@@ -415,26 +417,27 @@ protected function combine_campaigns_and_campaign_criterion_results( array $camp
+			$location_id         = $this->parse_geo_target_location_id( $geo_target_constant );
+
+			$google_helper = $this->container->get( GoogleHelper::class );
+			$country_code  = $google_helper->find_country_code_by_id( $location_id );
```
This function can return `null` if the `$location_id` is not a recognized country. It's probably not going to be possible to set an unrecognized country through our extension, but if they are already targeting a smaller region than a country (or set other locations in the dashboard) a null value will be added to the `targeted_locations` array.

ianlin replied:
Ah, my bad, should take more care on those cases. Updated in e2ac883, thanks a log!.

mikkamp commented on tests/Unit/API/Google/AdsCampaignTest.php:
```diff
@@ -41,12 +43,15 @@ public function setUp() {
+		$this->ad_group      = $this->createMock( AdsGroup::class );
+		$this->budget        = $this->createMock( AdsCampaignBudget::class );
+		$this->options       = $this->createMock( OptionsInterface::class );
+		$this->wc            = $this->createMock( WC::class );
+		$this->google_helper = new GoogleHelper( $this->wc );
```
I would add the `google_helper` as an `@property` in the class doc block. We don't really use `wc` as a property so I'd personally just pass it directly in the constructor.

```suggestion
		$this->google_helper = new GoogleHelper( $this->createMock( WC::class ) );
```

ianlin replied:
Thanks! This is also updated in 0247998.

mikkamp commented on tests/Unit/API/Google/AdsCampaignTest.php:
```diff
@@ -66,30 +71,14 @@ public function test_get_campaigns() {
 				'campaign_id'         => self::TEST_CAMPAIGN_ID,
 				'geo_target_constant' => 'geoTargetConstants/2158',
 			],
-			[
-				'campaign_id'         => self::TEST_CAMPAIGN_ID,
-				'geo_target_constant' => '',
-			],
-			[
-				'campaign_id'         => self::TEST_CAMPAIGN_ID,
-				'geo_target_constant' => null,
-			],
```
I can see these aren't relevant anymore because they will be skipped in a query. But what about other invalid locations that we expect to skip or throw an exception for:

```
geoTargetConstants/123456789
foobar/2158
```

ianlin replied:
Thanks for it, added a test case for it in c7436c8.

mikkamp commented on src/API/Google/AdsCampaign.php:
```diff
@@ -481,4 +484,21 @@ protected function parse_campaign_id( string $name ): int {
+			$parts = GeoTargetConstantServiceClient::parseName( $geo_target_constant );
+			return absint( $parts['criterion_id'] );
+		} catch ( ValidationException $e ) {
+			throw new Exception( __( 'Invalid geo target location ID', 'google-listings-and-ads' ) );
```
Just double checking if this is intentional or if you think it's better to skip the location. Since there is no try/catch around when we call this function, one bad location will prevent any campaigns from being retrieved.
That might be intentional though since the Google API shouldn't be returning invalid locations (in that case it's good to leave as is).

ianlin replied:
Yeah, it's intentional but skipping the location also makes sense to me 🤔 I kept it as is and added a test case to catch the exception in c7436c8 per your suggestion.

mikkamp reviewed (commented):
The PR is looking good and when testing I get the results I would expect with the country codes:

```json
{
	"id": 1234567890,
	"name": "Campaign 2022-01-01 12:00:00",
	"status": "enabled",
	"amount": 10,
	"country": "US",
	"targeted_locations": [
		"IE",
		"NL"
	]
}
```

I just left some thoughts on different things I noticed when going through the code changes.

ianlin committed:
Remove unused InvalidQuery class

ianlin committed:
Make GoogleHelper as a direct dependency for AdsCampaign class

Also changed:

- Pass wc to GoogleHelper's constructor directly rather than defind
	$this->wc property in AdsCampaignTest's setUp method.
- Add google_helper as a @property in AdsCampaignTest's class doc
	block.

ianlin committed:
Handle the case when country code can't be found by location ID

ianlin committed:
Add a test when invalid location IDs are returned from Google

ianlin commented:
Hey @mikkamp, thanks again for the review. I've updated the codes based on your suggestions and it's ready to be reviewed again.

mikkamp commented on src/API/Google/AdsCampaign.php:
```diff
@@ -71,10 +72,12 @@ class AdsCampaign implements ContainerAwareInterface, OptionsAwareInterface {
+	public function __construct( GoogleAdsClient $client, AdsCampaignBudget $budget, GoogleHelper $google_helper ) {
+		$this->client        = $client;
+		$this->budget        = $budget;
+		$this->google_helper = $google_helper;
```
Just missing to define this as a class parameter (like is done for the other two), otherwise it will default to public, but we want it to be protected.

ianlin replied:
👀 👀 Thanks! Added in 0811070.

mikkamp commented on src/API/Google/AdsCampaign.php:
```diff
@@ -415,25 +418,26 @@ protected function combine_campaigns_and_campaign_criterion_results( array $camp
+			$location_id         = $this->parse_geo_target_location_id( $geo_target_constant );
+			$country_code        = $this->google_helper->find_country_code_by_id( $location_id );
+
+			if ( isset( $country_code ) ) {
```
Just personal preference: `isset` is technically the correct way to identify it isn't `null`. Although the function definition doesn't guarantee that it can't return an empty string `''`. So I would have checked for both:

```suggestion
			if ( $country_code ) {
```

ianlin replied:
Good point, thanks! Updated in 46e5f7c.

mikkamp reviewed (approved):
Thanks for all the changes. I'm still getting the correct results when testing.

This looks good to ship, I've just left two small comments.

ianlin committed:
Define google_helper as protected class parameter in AdsCampaigns class

ianlin committed:
Check country_code using boolean comparison instead of isset

Since country_code could return both null or ''.
